# OBJECTIVE :

Understand what makes people interested in each other to go on a second date.  

# Notes before starting
1. There are waves in which some attributes had to be normalized
2. it has been noticed that there are some candidates who entered wrong values because their values added up to more than 100%, they were discarded
3. Differentiate  between a match and a going on date and particpiant

# 1) Imports

## 1.a) Import Libraries

In [2]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## 1.b) Import File

In [3]:
# Read the file and watch out for the encoding type
# choose the option that suits you , ignore the first becasue it is a mounted drive

#df = pd.read_csv("/content/drive/MyDrive/FullStack_projects/Speed_Dating/Speed+Dating+Data.csv", encoding = "ISO-8859-1")

df = pd.read_csv("Speed+Dating+Data.csv")

# if you don't have the file you can download it directly executing the link below:
# Uncomment the line below
# df = pd.read_csv("https://full-stack-assets.s3.eu-west-3.amazonaws.com/M03-EDA/Speed+Dating+Data.csv", encoding = "ISO-8859-1")



# 2) EDA

In [4]:
# setting display options

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

As a start is there any missing values concerning the 'match or not' flag and the gender?

In [5]:
# Count is missing labels , i.e. match results
print(f"number of missing labels match results : {df['match'].isna().sum()}")
print(f"number of missing gender : {df['gender'].isna().sum()}")

number of missing labels match results : 0
number of missing gender : 0


good we can rely on them , how about the dataset shape?

In [6]:
# Number of rows & columns:
print(f"number of rows : {df.shape[0]}")
print(f"number of columns : {df.shape[1]}")

number of rows : 8378
number of columns : 195


That's a huge number of columns 😀 we have to tackle these useful information

How does the data look like:

In [7]:
df.head()

iid   id  gender  idg  condtn  wave  round  position  positin1  order  \
0    1  1.0       0    1       1     1     10         7       NaN      4   
1    1  1.0       0    1       1     1     10         7       NaN      3   
2    1  1.0       0    1       1     1     10         7       NaN     10   
3    1  1.0       0    1       1     1     10         7       NaN      5   
4    1  1.0       0    1       1     1     10         7       NaN      7   

   partner   pid  match  int_corr  samerace  age_o  race_o  pf_o_att  \
0        1  11.0      0      0.14         0   27.0     2.0      35.0   
1        2  12.0      0      0.54         0   22.0     2.0      60.0   
2        3  13.0      1      0.16         1   22.0     4.0      19.0   
3        4  14.0      1      0.61         0   23.0     2.0      30.0   
4        5  15.0      1      0.21         0   24.0     3.0      30.0   

   pf_o_sin  pf_o_int  pf_o_fun  pf_o_amb  pf_o_sha  dec_o  attr_o  sinc_o  \
0      20.0      20.0      20.0       0.0       5.0      0     6.0     8.0   
1       0.0       0.0      40.0       0.0       0.0      0     7.0     8.0   
2      18.0      19.0      18.0      14.0      12.0      1    10.0    10.0   
3       5.0      15.0      40.0       5.0       5.0      1     7.0     8.0   
4      10.0      20.0      10.0      10.0      20.0      1     8.0     7.0   

   intel_o  fun_o  amb_o  shar_o  like_o  prob_o  met_o   age field  field_cd  \
0      8.0    8.0    8.0     6.0     7.0     4.0    2.0  21.0   Law       1.0   
1     10.0    7.0    7.0     5.0     8.0     4.0    2.0  21.0   Law       1.0   
2     10.0   10.0   10.0    10.0    10.0    10.0    1.0  21.0   Law       1.0   
3      9.0    8.0    9.0     8.0     7.0     7.0    2.0  21.0   Law       1.0   
4      9.0    6.0    9.0     7.0     8.0     6.0    2.0  21.0   Law       1.0   

  undergra mn_sat tuition  race  imprace  imprelig     from zipcode  \
0      NaN    NaN     NaN   4.0      2.0       4.0  Chicago  60,521   
1      NaN    NaN     NaN   4.0      2.0       4.0  Chicago  60,521   
2      NaN    NaN     NaN   4.0      2.0       4.0  Chicago  60,521   
3      NaN    NaN     NaN   4.0      2.0       4.0  Chicago  60,521   
4      NaN    NaN     NaN   4.0      2.0       4.0  Chicago  60,521   

      income  goal  date  go_out  career  career_c  sports  tvsports  \
0  69,487.00   2.0   7.0     1.0  lawyer       NaN     9.0       2.0   
1  69,487.00   2.0   7.0     1.0  lawyer       NaN     9.0       2.0   
2  69,487.00   2.0   7.0     1.0  lawyer       NaN     9.0       2.0   
3  69,487.00   2.0   7.0     1.0  lawyer       NaN     9.0       2.0   
4  69,487.00   2.0   7.0     1.0  lawyer       NaN     9.0       2.0   

   exercise  dining  museums  art  hiking  gaming  clubbing  reading   tv  \
0       8.0     9.0      1.0  1.0     5.0     1.0       5.0      6.0  9.0   
1       8.0     9.0      1.0  1.0     5.0     1.0       5.0      6.0  9.0   
2       8.0     9.0      1.0  1.0     5.0     1.0       5.0      6.0  9.0   
3       8.0     9.0      1.0  1.0     5.0     1.0       5.0      6.0  9.0   
4       8.0     9.0      1.0  1.0     5.0     1.0       5.0      6.0  9.0   

   theater  movies  concerts  music  shopping  yoga  exphappy  expnum  \
0      1.0    10.0      10.0    9.0       8.0   1.0       3.0     2.0   
1      1.0    10.0      10.0    9.0       8.0   1.0       3.0     2.0   
2      1.0    10.0      10.0    9.0       8.0   1.0       3.0     2.0   
3      1.0    10.0      10.0    9.0       8.0   1.0       3.0     2.0   
4      1.0    10.0      10.0    9.0       8.0   1.0       3.0     2.0   

   attr1_1  sinc1_1  intel1_1  fun1_1  amb1_1  shar1_1  attr4_1  sinc4_1  \
0     15.0     20.0      20.0    15.0    15.0     15.0      NaN      NaN   
1     15.0     20.0      20.0    15.0    15.0     15.0      NaN      NaN   
2     15.0     20.0      20.0    15.0    15.0     15.0      NaN      NaN   
3     15.0     20.0      20.0    15.0    15.0     15.0      NaN      NaN   
4     15.0     

An overview descriptionof the dataset we are dealing with:

In [8]:
df.describe()

iid           id       gender          idg       condtn  \
count  8378.000000  8377.000000  8378.000000  8378.000000  8378.000000   
mean    283.675937     8.960248     0.500597    17.327166     1.828837   
std     158.583367     5.491329     0.500029    10.940735     0.376673   
min       1.000000     1.000000     0.000000     1.000000     1.000000   
25%     154.000000     4.000000     0.000000     8.000000     2.000000   
50%     281.000000     8.000000     1.000000    16.000000     2.000000   
75%     407.000000    13.000000     1.000000    26.000000     2.000000   
max     552.000000    22.000000     1.000000    44.000000     2.000000   

              wave        round     position     positin1        order  \
count  8378.000000  8378.000000  8378.000000  6532.000000  8378.000000   
mean     11.350919    16.872046     9.042731     9.295775     8.927668   
std       5.995903     4.358458     5.514939     5.650199     5.477009   
min       1.000000     5.000000     1.000000     1.000000     1.000000   
25%       7.000000    14.000000     4.000000     4.000000     4.000000   
50%      11.000000    18.000000     8.000000     9.000000     8.000000   
75%      15.000000    20.000000    13.000000    14.000000    13.000000   
max      21.000000    22.000000    22.000000    22.000000    22.000000   

           partner          pid        match     int_corr     samerace  \
count  8378.000000  8368.000000  8378.000000  8220.000000  8378.000000   
mean      8.963595   283.863767     0.164717     0.196010     0.395799   
std       5.491068   158.584899     0.370947     0.303539     0.489051   
min       1.000000     1.000000     0.000000    -0.830000     0.000000   
25%       4.000000   154.000000     0.000000    -0.020000     0.000000   
50%       8.000000   281.000000     0.000000     0.210000     0.000000   
75%      13.000000   408.000000     0.000000     0.430000     1.000000   
max      22.000000   552.000000     1.000000     0.910000     1.000000   

             age_o       race_o     pf_o_att     pf_o_sin     pf_o_int  \
count  8274.000000  8305.000000  8289.000000  8289.000000  8289.000000   
mean     26.364999     2.756653    22.495347    17.396867    20.270759   
std       3.563648     1.230689    12.569802     7.044003     6.782895   
min      18.000000     1.000000     0.000000     0.000000     0.000000   
25%      24.000000     2.000000    15.000000    15.000000    17.390000   
50%      26.000000     2.000000    20.000000    18.370000    20.000000   
75%      28.000000     4.000000    25.000000    20.000000    23.810000   
max      55.000000     6.000000   100.000000    60.000000    50.000000   

          pf_o_fun     pf_o_amb     pf_o_sha        dec_o       attr_o  \
count  8280.000000  8271.000000  8249.000000  8378.000000  8166.000000   
mean     17.459714    10.685375    11.845930     0.419551     6.190411   
std       6.085526     6.126544     6.362746     0.493515     1.950305   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%      15.000000     5.000000     9.520000     0.000000     5.000000   
50%      18.000000    10.000000    10.640000     0.000000     6.000000   
75%      20.000000    15.000000    16.000000     1.000000     8.000000   
max      50.000000    53.000000    30.000000     1.000000    10.500000   

            sinc_o      intel_o        fun_o        amb_o       shar_o  \
count  8091.000000  8072.000000  8018.000000  7656.000000  7302.000000   
mean      7.175256     7.369301     6.400599     6.778409     5.474870   
std       1.740575     1.550501     1.954078     1.794080     2.156163   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       6.000000     6.000000     5.000000     6.000000     4.000000   
50%       7.000000     7.000000     7.000000     7.000000     6.000000   
75%       8.000000     8.000000     8.000000     8.000000     7.000000   
max      10.000000    10.000000    11.000000    10.000000    10.000000   

            l

### 2.1) Dataset check list

According to the discriptions and after comparing:

pf0_att ====> normalized out of 100 %

attr_0  => participant evaluate the partner => scaled out of 10

attr  => Partner evaluate the particpant    => scaled out of 10

attr4_1 => => what the participant think => mixed => waves 6-9 needs normalizing ==> Done...

attr1_1 => looking for => all normalized

attr2_1 => opposit sex look for all normalized

attr3_1 => rate yourself => scaled out of 10 => need normalizing  ==> Done...

attr5_1 => scaled out of 10 => need normalizing depennding on comparison type


attr7_2 => looking for => normalized out 100%

attr1-2 => look for => all normalized

attr1_s => too many missing data

## 2.2) Normalizing

Normalizing the necessary attributes in waves from 6 to 9 tobecome out of 100

In [9]:
# Normalizing the necessary attributes that needs to be mormailzed in the waves from 6 to 9 tobecome out of 100
df['total4_1'] = df['attr4_1'] + df['sinc4_1'] + df['intel4_1'] + df['fun4_1'] + df['amb4_1'] + df['shar4_1']
df['total4_2'] = df['attr4_2'] + df['sinc4_2'] + df['intel4_2'] + df['fun4_2'] + df['amb4_2'] + df['shar4_2']
normalized_value = 0

def normalize(row):
  # filter the waves concerned
  if ((row['wave'] >  5) & (row['wave'] < 10)):

    normailzed_value = row[col]/row['total4_1']
    return normalized_value
  return row[col]   # retunr the original value if the condition is not met

# loop through the columns to normalize
cols = ['attr4_1','sinc4_1','intel4_1', 'fun4_1','amb4_1','shar4_1']
for col in cols:
  df[col] = df.apply(normalize, axis=1) # call the normalizing function
                                        # Don't forget axis = 1


# NOTE : No worries for empty values because they will be filtered out later on, during the calculation

In [10]:
df['attr4_1'].describe()

count    6494.000000
mean       24.302895
std        18.780153
min         0.000000
25%        10.000000
50%        25.000000
75%        35.000000
max        95.000000
Name: attr4_1, dtype: float64

Seems ok, let's proceed with the final normailzing

In [11]:
cols = ['attr4_2','sinc4_2','intel4_2', 'fun4_2','amb4_2','shar4_2']
for col in cols:
  df[col] = df.apply(normalize, axis=1)

In [12]:
df['attr4_2'].describe()

count    5939.000000
mean       24.031824
std        19.105127
min         0.000000
25%         0.000000
50%        25.000000
75%        40.000000
max       100.000000
Name: attr4_2, dtype: float64

# Approach:

We shall implement our investigation in two phases ,
1. The first is the general analysis which gives us an overview of the whole picture.
2. Detailed Analysis 🔍

  2.1 Attributes analysis in details which explains what's going in reality

  2.2 Highlighting some indicators

3. Summary

# A) General Statistic

Before starting I prefer to add a unique id for each candidate , this will be useful for future use, as we will see  


In [13]:
# I prefer to add a unique id for each candidate , this will be useful for future use,
# this unique id is comprised of the wave number, the candidate id number in this wave, and male or female

# assign unique id based on values of columns
def assign_unique_id(row):

  # extracting male or female
  if row['gender'] == 0:
    gender = 'F'
  else:
    gender = 'M'

  unique_id =  gender + str(row['wave']) + '_' + str(row['id'])
  return unique_id

# create the unique_id column
df['unique_id'] = df.apply(assign_unique_id, axis=1)  # don't forget (axis = 1) to differentiate between column or rows direction not rows

### A.1) Total number of participants

How many candidates do we have?

In [14]:
print('Total number of participants:')
len(df['unique_id'].unique())

Total number of participants:


552

we have 552 candidates,


### A.2) Number of female and male candidates _*per wave*_

In [15]:
# Choosing the columns of interest
df_temp = df.loc[:,['wave','gender', 'unique_id']]

# removing the duplicates becasue of interviews
df_temp = df_temp.drop_duplicates()

# Counting
group_by_wave = df_temp.groupby(['wave','gender'])['unique_id'].count()

# to beautify the output and to look like a data frame
group_by_wave = group_by_wave.reset_index()
display(group_by_wave)

wave  gender  unique_id
0      1       0         10
1      1       1         10
2      2       0         19
3      2       1         16
4      3       0         10
5      3       1         10
6      4       0         18
7      4       1         18
8      5       0          9
9      5       1         10
10     6       0          5
11     6       1          5
12     7       0         16
13     7       1         16
14     8       0         10
15     8       1         10
16     9       0         20
17     9       1         20
18    10       0          9
19    10       1          9
20    11       0         21
21    11       1         21
22    12       0         14
23    12       1         14
24    13       0         10
25    13       1          9
26    14       0         20
27    14       1         18
28    15       0         18
29    15       1         19
30    16       0          6
31    16       1          8
32    17       0         10
33    17       1         14
34    18       0          6
35    18       1          6
36    19       0         15
37    19       1         15
38    20       0          6
39    20       1          7
40    21       0         22
41    21       1         23

the table above illustrates for each wave how many female participant and how many are males. The results are exactly the same as the word document

## The distribution of the participants

In [112]:
# estalishing a table with the conerned features
particpants_features = ['unique_id','age', 'field', 'undergra', 'mn_sat','tuition','race', 'imprace', 'imprelig', 'from' , 'zipcode',	'income', 'goal',	'date'	,'go_out','career', 'gender', 'date_3','length']

df_participants = df.loc[:, particpants_features].copy()
df_participants = df_participants.drop_duplicates()

print(f'The number of rows is {df_participants.shape[0]}')


The number of rows is 552


## A.4) Age distibution

In [17]:
df_participants['student'] = df_participants['undergra'].apply(lambda x: 0 if pd.notnull(x) else 1)

In [18]:
df_participants['student'].value_counts()

0    316
1    236
Name: student, dtype: int64

In [19]:
fig = px.histogram(df_participants , x='age' ,title="Age distibution of applicants")
fig.show()

In [20]:
fig = px.histogram(df_participants , x='age' ,
                   color='date_3',
                   title='Age distribution with second date',
                 )
fig.data[0].name = "Didn't go on date"
fig.data[1].name = 'Went on a date with the match'
fig.show()

It is important how to read this chart, first of all 52% did not fill if they went on a second date or not. From the data we have the age is not very decisive in having a second date or not becasue in the end of the day it takes almost the same distribution of the number of applicants, i.e. participants

## A.5 ) Race distribution


In [21]:
mapping_race = {1: 'Black/African American', 2: 'European/Caucasian-American', 3: 'Latino/Hispanic American', 4: 'Asian/Pacific Islander/Asian-American', 5: 'Native American', 6: 'Other'}
df_participants['race'] = df_participants['race'].replace(mapping_race)

In [22]:
fig = px.pie(df_participants , names='race' ,title="Race distibution of applicants")

fig.show()

# B) Match Cases Analyses

When we say match , it means that both sides liked each other as a start.
we shall analyse the factors that might played a role in making or achiveing this match

## B.1) The role of the race:
Did the **race** play a role in increasing the number of matches.
we will count on a flag called "samerace" meaning are they of the same race

In [23]:
# could you please make sure that there is no null
df['samerace'].isna().sum()

0

In [24]:
# replace the number values with text
df['samerace'] = df['samerace'].apply(lambda x :  'yes' if x == 1 else 'no')


In [25]:
df.groupby('match')['samerace'].value_counts()

match  samerace
0      no          4248
       yes         2750
1      no           814
       yes          566
Name: samerace, dtype: int64

The table above shows that within the match cases the percentage of applicants that had the same race is less than 50%

**Conclusion** the race does not play a role in having a match

the chart below shows the same conclusion

In [26]:
fig = px.histogram(df,x='match' , color='samerace', title="Race similarity distribution in both categories (Match/No match)")
fig.show()

**Conclusion** the race does not play a role in having a match

## B.2) Previous knowledge of each other

Could it be that if they met before , chances are higher?

In [27]:
# could you please make sure that there is no null
df['met'].isna().sum()/df.shape[0] * 100

4.4760085939365

4% of missing values no big deal

In [28]:
mask = df['match'] == 1
print(f"The total number of matches was {df.loc[mask,:].shape[0]}")

The total number of matches was 1380


In [29]:
# we don't care how many times they met, i need to know if they have met before or not
df['met_before'] = df['met'].apply(lambda x : 'yes' if x > 0 else 'no')

In [30]:
df.groupby(['met_before','match'])['match'].count()

met_before  match
no          0        3788
            1         634
yes         0        3210
            1         746
Name: match, dtype: int64

Here the distribution total is equal to the total line of matches minus the missing values of the 'met' column. the result is accurate. Let's have a look at the chart

In [31]:
fig = px.histogram(df,x='met_before' , color='match', title="Chances of getting a match if they have met before")
fig.show()

**Conclusion 2:** The distribution illustrates that even if they met before doesn't play a role either.

### Distinguishing dataset

To distinguish between the match and not matched cases' I have split the mtach cases and no_match.

In [32]:
# to analyze the cases match
mask = df['match'] == 1
df_match = df.loc[mask,:].copy()

mask = df['match'] == 0
df_no_match = df.loc[mask,:].copy()


In [33]:
print(f'number of matched cases : {df_match.shape[0]}')
print(f'number of no matched cases : {df_no_match.shape[0]}')

number of matched cases : 1380
number of no matched cases : 6998


# C) Attributes Analysis

## C.1) Preparations

##### Attributes' Lists created

It would be wise to organize the features attributes in lists according to their types and timings.

This is to ease up work and to facilite the treatment much easily

In [34]:
# All attributes are grouped together to facilitate the code and for logical understanding

attrib_preffered_by_partener = ['pf_o_att','pf_o_sin','pf_o_int', 'pf_o_fun','pf_o_amb','pf_o_sha']  #partner’s stated preference at Time 1 (attr1_1) for all 6 attributes
attrib_rated_by_partner = ['attr_o','sinc_o','intel_o', 'fun_o','amb_o'] # rating by partner the night of the event, for all 6 attributes
attrib_looking_for = ['attr1_1','sinc1_1','intel1_1', 'fun1_1','amb1_1','shar1_1'] # what you look for in the opposite sex.
attrib_rate_yourself = ['attr3_1','sinc3_1','intel3_1', 'fun3_1','amb3_1'] # How do you think you measure up?
attrib_score_card_evaluation = ['attr','sinc','intel', 'fun','amb','shar']  # Filled out by subjects after each "date" during the event
# the last one above is how this particpant evaluated his/her partner

# these were asked aging during halftime
attrib_looking_for_halftime = ['attr1_s','sinc1_s','intel1_s', 'fun1_s','amb1_s','shar1_s'] # what you look for in the opposite sex.
attrib_rate_yourself_halftime = ['attr3_s','sinc3_s','intel3_s', 'fun3_s','amb3_s']


# these are after one day of the date
attrib_looking_for_2nd_day = ['attr1_2','sinc1_2','intel1_2', 'fun1_2','amb1_2','shar1_2'] # what you look for in the opposite sex.
attrib_rate_yourself_2nd_day = ['attr3_2','sinc3_2','intel3_2', 'fun3_2','amb3_2'] # How do you think you measure up?

# after 3 weeks
attrib_looking_for_after3weeks = ['attr1_3','sinc1_3','intel1_3', 'fun1_3','amb1_3','shar1_3'] # what you look for in the opposite sex.
attrib_more_important_after3weeks = ['attr7_3','sinc7_3','intel7_3', 'fun7_3','amb7_3','shar7_3'] # attributes that were more important in your decisions
attrib_rate_yourselfafter3weeks = ['attr3_3','sinc3_3','intel3_3', 'fun3_3','amb3_3'] # Please rate your opinion of your own attributes

# this is just an opinion and does not reflect reality
attrib_fellow_look_for_think = ['attr4_1','sinc4_1','intel4_1', 'fun4_1','amb4_1','shar4_1'] # what you think MOST of your fellow men/women look for in the opposite sex.
attrib_fellow_look_for_think_2nd_day = ['attr4_2','sinc4_2','intel4_2', 'fun4_2','amb4_2','shar4_2'] # what you think MOST of your fellow men/women look for in the opposite sex , second day survey

attrib_oppositSex_look_for_think = ['attr2_1','sinc2_1','intel2_1', 'fun2_1','amb2_1','shar2_1']# What do you think the opposite sex looks for in a date?
attrib_oppositSex_look_for_think_2nd_day = ['attr2_2','sinc2_2','intel2_2', 'fun2_2','amb2_2','shar2_2']# What do you think the opposite sex looks for in a date?

attrib_others_percieve_you_think = ['attr5_1','sinc5_1','intel5_1', 'fun5_1','amb5_1'] # how do you think others perceive you?
attrib_others_percieve_you_think_2nd_day = ['attr5_2','sinc5_2','intel5_2', 'fun5_2','amb5_2'] # how do you think others perceive you?




#### Notes

1. Due to the nature this dataset was organized when filtering the profiles, there will be duplicate lines. These duplicate lines need to be removed otherwise they will faslify the results and make them biased

2. **We** can see that some candidates in some cases did NOT fill the whole percentages during the questionnaire,

 **Important** this will falsify the results, need to eliminate and drop these rows that have incomplete attributes during the calculations

Hereby is an example of empty fields for the column ' preffered attributes' for instance

In [35]:
for column in attrib_preffered_by_partener:
  print(column)
  print(df_match[column].isna().sum())

pf_o_att
15
pf_o_sin
15
pf_o_int
15
pf_o_fun
17
pf_o_amb
20
pf_o_sha
20


### C.2) ATTRIBUTES OBSERVER

Establishing a petite dataframe for the results and *charts* especially designed to be use dby plotly

In [149]:
# two dataframes one that looks like a table for reporting and troubleshooting
df_attributes = pd.DataFrame(columns = ['gender','status','Atractive','Sincere','Intelligent','Fun','Ambitious','Shared_interest','flag'])

# the second is designed to be used by plotly
df_graph = pd.DataFrame(columns = ['gender','attribute','value','status'])

In [150]:
# This function will calcualte all attributes and produce a table report
# And another graph dataset especially to use very simply plotly features


def calculate_attributes(scope_to_cover,scope_name):
  print('Report started')
  for scope, name in zip(scope_to_cover,scope_name):

    # 1) preparing the data
    needed_columns = scope.copy()     # extracting the needed attributes
    needed_columns = needed_columns + ['gender','unique_id']  # adding gender and unique_id
    df_lab = df.loc[:, needed_columns].copy()  # masking all the mentioned
    # try the above with df_match and df_no_match they all produce the same results

    print(f'Calculating for attribute "{name}":')
    print(f' dataframe shape is {df_lab.shape}')
    print()

    print(' dropping rows with empty fields...' )
    df_lab = df_lab.dropna()

    print(' Data shape after dropping the empty fields:' ,df_lab.shape)
    print()

    ## === Attention === ###
    ## There are duplicate lines becasue these attributes are repeated for each interview
    print(' Checking duplicates and dropping them..')
    df_lab = df_lab.drop_duplicates()
    print('After dropping', df_lab.shape)
    print()

    # 2) Calculate
    grouped_res = df_lab.groupby('gender')[scope].mean()
    display(grouped_res)

    # 3) Inserting into report table format
    # 3.1) define a new item to insert in attribute_dataframe for reporting
    # why? because we want it like offical report
    for i in range(2):
      new_item = {'gender': i,
                  'status': name,
                  'Atractive': grouped_res.iloc[i,0],
                  'Sincere':grouped_res.iloc[i,1],
                  'Intelligent':grouped_res.iloc[i,2],
                  'Fun':grouped_res.iloc[i,3],
                  'Ambitious':grouped_res.iloc[i,4],
                  'Shared_interest':grouped_res.iloc[i,5],
                  'flag': 0 }

      # 3.2 )define a pointer at the end of the df_attribute to insert data
      index_df_attributes = df_attributes.shape[0]

      # 3.3) Insert data
      df_attributes.loc[index_df_attributes] = new_item

      # 4 ) graph table to benefit from all plotly features
      attribute_names = ['Attractive','Sincere','Intelligent','Fun','Ambitious','Shared_interest']
      for attribute_index in range(5):
        graph_entity = {'gender':i,
                      'attribute': attribute_names[attribute_index],
                        'value': grouped_res.iloc[i,attribute_index],
                        'status': name
                        }

        # define graph dataset indicator to insert
        index_df_graph = df_graph.shape[0]

        # Insert
        df_graph.loc[index_df_graph] = graph_entity





## C.3) Preferred Attributes analysis

What are the attributes that both genders aiming at and which are most important?

Report:

In [151]:
# The overall attributes that the candidates are looking for
scope_to_cover =[attrib_looking_for,
                 attrib_looking_for_2nd_day,
                 attrib_looking_for_after3weeks,
                 attrib_more_important_after3weeks
                 ]

scope_name = ['attrib_looking_for',
                 'attrib_looking_for_2nd_day',
                 'attrib_looking_for_after3weeks',
                 'attrib_more_important_in_the_end'
                 ]

calculate_attributes(scope_to_cover,scope_name)

Report started
Calculating for attribute "attrib_looking_for":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (8257, 8)

 Checking duplicates and dropping them..
After dropping (542, 8)



attr1_1    sinc1_1   intel1_1     fun1_1     amb1_1    shar1_1
gender                                                                  
0       17.938358  18.252910  20.937313  17.326343  12.828993  12.697836
1       27.165766  16.329891  19.403431  17.582591   8.791752  10.942190

Calculating for attribute "attrib_looking_for_2nd_day":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (7445, 8)

 Checking duplicates and dropping them..
After dropping (485, 8)



attr1_2    sinc1_2   intel1_2     fun1_2     amb1_2    shar1_2
gender                                                                  
0       21.862996  16.562194  19.114388  17.394895  11.495148  13.916962
1       30.287500  15.143226  16.744395  17.705161   8.500968  11.485766

Calculating for attribute "attrib_looking_for_after3weeks":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (3974, 8)

 Checking duplicates and dropping them..
After dropping (264, 8)



attr1_3    sinc1_3   intel1_3     fun1_3     amb1_3    shar1_3
gender                                                                  
0       20.565685  17.313767  19.846712  15.851301  12.294178  14.129589
1       29.528220  15.844746  18.956186  16.332797   8.899915  10.777881

Calculating for attribute "attrib_more_important_in_the_end":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (2016, 8)

 Checking duplicates and dropping them..
After dropping (132, 8)



attr7_3    sinc7_3   intel7_3     fun7_3    amb7_3    shar7_3
gender                                                                 
0       26.958904  18.123288  16.452055  15.054795  9.465753  14.082192
1       37.305085  13.118644  17.220339  17.067797  5.796610   9.661017

### Preferred Atributes "Evolution with time"

In [152]:
fig= px.bar(df_graph, x='status', color='attribute', y='value', title="Prefferred Attributes per gender & Evolution in time", barmode='group', facet_row='gender')
fig.show()

### **Conclusion & Observations**

**The females:**
1. In the begining, the females were interested mostly in intelligence, but then this changed as the attractiveness came as the highest priority. That was a major change
2. Always the attribues Ambitious and funny were the least desired

**The Males**
2. As for the males gender, since the begining it was Attractiveness which had the highest priority and moreover increased through the phases.
3. Also Ambitious was the least attribute, after that comes sincerety

### Male Vs. Female prefered Attributes

In [98]:
fig= px.bar(df_graph, x='gender', color='attribute', y='value', title="Male Vs. Female prefered Attributes", barmode='group',facet_col='status')
fig.show()

The chart above illsutrates the same remarks and conlusions we have made in the cell above

## C.4) Perception Analysis

What does the majority of each gender think that the other gender is interested in? and what do they think their fellows from the same gender are interested in ?

In [99]:
# The overall attributes that the candidates are looking for
scope_to_cover =[attrib_fellow_look_for_think,
                 attrib_fellow_look_for_think_2nd_day,
                 attrib_oppositSex_look_for_think,
                 attrib_oppositSex_look_for_think_2nd_day,

                 ]

scope_name = ['attrib_fellow_look_for_think',
                 'attrib_fellow_look_for_think_2nd_day',
                 'attrib_oppositSex_look_for_think',
                 'attrib_oppositSex_look_for_think_2nd_day',
                 ]

calculate_attributes(scope_to_cover,scope_name)

Report started
Calculating for attribute "attrib_fellow_look_for_think":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (6472, 8)

 Checking duplicates and dropping them..
After dropping (418, 8)



attr4_1   sinc4_1   intel4_1     fun4_1     amb4_1   shar4_1
gender                                                                
0       22.247573  9.407767  11.097087  13.053398  10.033981  9.388350
1       26.448113  9.113208  10.462264  14.155660   6.353774  9.235849

Calculating for attribute "attrib_fellow_look_for_think_2nd_day":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (5939, 8)

 Checking duplicates and dropping them..
After dropping (380, 8)



attr4_2    sinc4_2   intel4_2     fun4_2    amb4_2   shar4_2
gender                                                                
0       21.352941  10.438503  10.823529  12.000000  8.518717  9.513369
1       26.580311   9.082902   9.264249  13.455959  6.616580  8.922280

Calculating for attribute "attrib_oppositSex_look_for_think":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (8289, 8)

 Checking duplicates and dropping them..
After dropping (544, 8)



attr2_1    sinc2_1   intel2_1     fun2_1     amb2_1    shar2_1
gender                                                                  
0       35.600632  11.284535  12.478439  19.051636   9.114387  12.626766
1       24.975855  15.053527  16.294800  18.087891  14.305273  11.196473

Calculating for attribute "attrib_oppositSex_look_for_think_2nd_day":
 dataframe shape is (8378, 8)

 dropping rows with empty fields...
 Data shape after dropping the empty fields: (5775, 8)

 Checking duplicates and dropping them..
After dropping (369, 8)



attr2_2    sinc2_2   intel2_2     fun2_2     amb2_2    shar2_2
gender                                                                  
0       32.943500  12.094556  12.492944  18.768000   9.672611  13.767556
1       25.425238  15.850423  15.593757  17.228836  14.229312  11.842540

In [100]:
# adding a flag for each row that contains the perception context i.e. the word think attribute
df_graph['discrminiator'] = df_graph['status'].apply(lambda x : 'perception' if 'think' in x else '')

In [101]:
mask = df_graph['discrminiator'] == 'perception'
temp = df_graph.loc[mask]
fig= px.bar(temp, x='status', color='attribute', y='value', title="Perception Evolution with time", barmode='group', facet_row='gender')
fig.show()

### **Conclusion & Observations**

This time it is not what people prefer.. No , it's what they think that others prefer.

1. They all believed that all is searching for attractiveness.

despite the fact if we remember from above , that in reality in the first phase from the females prespective , intelligence was a bit higher than the others.

2. Secondly, always funny was the second place , unlike the stats of what people wanted.

after half time note:
Too many candidates did not fill the fileds in the half time

## C.5) Rating analysis
There are two ratings
First is the personal rating of the particpant to himself

Second the the ratings of the partners who met with him.

How compatible are these results together, are the particpants able to evaluate themselves right? are the partners evaluation accurate?

bearing in mind that **I CHOSE ON PURPOSE TO ANALYZE THE CASES WHERE A MATCH OCCURRED.**

In [103]:
## APOLOGIES THIS PART ONLY OF CODE IS NOT DRY, BECAUSE LIMITED IN TIME
r1 = ['attr3_1','sinc3_1','intel3_1', 'fun3_1','amb3_1']
r2 = ['attr3_2','sinc3_2','intel3_2', 'fun3_2','amb3_2']
r3 = ['attr_o','sinc_o','intel_o', 'fun_o','amb_o']

df_ratings = pd.DataFrame(
    columns=['unique_id','gender', 'attribute', 'value','status']
)

for col in r1:

  needed_columns = ['unique_id','gender',col,'partner']
  bring = df_match.loc[:,needed_columns]

  # Get the existing column names
  column_names = bring.columns.tolist()

  # Rename the third column
  column_names[2] = 'value'

  bring.columns = column_names

  bring['status'] = 'evaluate_yourself'
  bring['attribute'] = col

  df_ratings = pd.concat([df_ratings, bring])
#############################
for col in r2:

  needed_columns = ['unique_id','gender',col,'partner']
  bring = df_match.loc[:,needed_columns]

  # Get the existing column names
  column_names = bring.columns.tolist()

  # Rename the third column
  column_names[2] = 'value'

  bring.columns = column_names

  bring['status'] = 'evaluate_yourself_after3weeks'
  bring['attribute'] = col

  df_ratings = pd.concat([df_ratings, bring])
###############################

for col in r3:

  needed_columns = ['unique_id','gender',col,'partner']
  bring = df_match.loc[:,needed_columns]

  # Get the existing column names
  column_names = bring.columns.tolist()

  # Rename the third column
  column_names[2] = 'value'

  bring.columns = column_names

  bring['status'] = 'rate_by_partner'
  bring['attribute'] = col

  df_ratings = pd.concat([df_ratings, bring])

def condition(x):
    if 'attr' in x:
        return "Attractive"
    elif 'sinc' in x :
        return "Sincere"
    elif 'intel' in x:
        return 'Intelligent'
    elif 'fun' in x:
        return 'Funny'
    elif 'amb' in x:
        return 'Ambitious'
    else :
        return 'NA'

# Applying the conditions
df_ratings['attribute'] = df_ratings['attribute'].apply(condition)

In [125]:
mask = df_ratings['unique_id'].str.contains("F1.1")
#mask = (df_ratings['unique_id'] == 'F1.1') | (df_ratings['unique_id'] == 'F1.2')

#mask = df_ratings['unique_id'] == 'F1_1.0'
df_chart = df_ratings.loc[mask,:]
fig= px.bar(df_chart, x='status', color='attribute', y='value', title="Perception Evolution with time", barmode='group',
            facet_row='unique_id',
            facet_col='partner')
fig.show()

## OBSERVATIONS & CONCLUSIONS
1. This particpant for example evaluated himself dramatically in a different manner between the begining and after three weeks.

2. On the other hand
was evaluated by four partners where a match occurred for all these interviews had different ratings.

Apparently , these participants are unable to rank themselves right.

on the other hand the partners made different ratings for the same person despite a match occurred

**most partners found him attractive except the particpant himself 😀**

Here is anothe example of a female evaluated by a male

In [123]:
#mask = df_ratings['unique_id'].str.contains("F1.1","F1.2")
mask = (df_ratings['unique_id'] == 'M21_5.0')

#df_ratings['unique_id'] == 'F1_1.0'
df_chart = df_ratings.loc[mask,:]
fig = px.bar(df_chart, x='status', color='attribute', y='value', title="Self Evaluation Vs. partner evaluation", barmode='group',
            facet_row='unique_id',
            facet_col='partner')
fig.show()

Same conclusion as above regaring the particapnt and her matched partners, which leads me to the upcomming question

### Interview length
Is the length of the interview a problem?

In [135]:
print('========== Length Evaluation==========')
print('1. The length distribtion')
print()
print(df_participants.groupby(['length'])['unique_id'].count())

print()

print('2. The length and date success distribtion')
print()
print(df_participants.groupby(['length','date_3'])['unique_id'].count())

print('1 is too short, 2 is very long, 3 is adequate')

========== Length Evaluation==========
1. The length distribtion

length
1.0    270
2.0     14
3.0    202
Name: unique_id, dtype: int64

2. The length and date success distribtion

length  date_3
1.0     0.0       83
        1.0       45
2.0     0.0        8
        1.0        2
3.0     0.0       67
        1.0       42
Name: unique_id, dtype: int64
1 is too short, 2 is very long, 3 is adequate


This is an astonishing discovery, both genders are are not confident how people see them, or maybe they are but for sure noto viewed as attractive

### Meeting order?
In terms of getting a second date, is it better to be someone's first speed date of the night or their last?

In [139]:
# to prevent duplicated values filter onone gender, +> the same results

mask = df['gender'] == 0

df_lab = df_match.loc[mask,['unique_id','order','wave','id', 'pid' ,'position']]
print(df_lab.shape)

# drop duplicates otherwise the results will be biased
df_lab = df_lab.drop_duplicates()

print(type(df_lab))
print(df_lab.shape)

fig = px.histogram(df_lab, x='order'  , title="Is it better to be someone's first speed date of the night or their last? ")
fig.show()

(690, 6)
<class 'pandas.core.frame.DataFrame'>
(690, 6)


## Lucky table

Are shared interests more important than a shared racial background?


In [140]:
fig = px.histogram(df_lab, x='position'  , title="Is there a lucky table position?")
fig.show()

## Shared interest

In [144]:
fig = px.histogram(df_match, x='int_corr', histnorm='percent' ,title='Interest rate distribution for the matched casest')
fig.show()
fig = px.histogram(df_no_match, x='int_corr',histnorm='percent',title='Interest rate distribution for the non matched cases')
fig.show()

The macthed data graph is skewed to the right indicating that shared interest plays a role in the improving the match rate

# THE END OF THE REPORT
the rest is just a lab

# Appendix & scratch (please ignore)

The majority of particpants are from the USA no need to waste time in categorizing the states , concentrate on more important

In [ ]:
#group_origin = df_participants.groupby('from')['from'].count()
df_participants['from'].value_counts()

In [ ]:
# old way for counting gender distribution